pipeline/scripts/ci/upload_compose.py

```diff
@@ -0,0 +1,151 @@
+[{{ data.nvr }}-{{ repo }}]
+name = {{ data.nvr }}-{{ repo}}
+baseurl = {{ data.base_url }}/compose/{{ repo }}/$basearch/os
+enabled = 0
+gpgcheck = 0
+
+{% endfor %}
+"""
+
+usage = """Upload Compose.
+
+This script pulls the development RPMs available in the given base_url and uploads an
+archive of it using the provided object_name to the mentioned bucket.
+
+In the context of CephCI,
+    bucket_name     is the name of the RHCS build
+    object_name     is the ceph version
+    base_url        the base url containing the RPMs to be uploaded
+
+  Example:
+    python upload_compose.py rhceph-4.2-rhel-7 ceph-14.6.2-113 http://download.rh.com/x
+
+Usage:
+  upload_compose.py <bucket_name> <object_name> <base_url>
+  upload_compose.py -h | --help
+
+Options:
+  -h --help     Show this screen
+
+"""
+
+
+def create_repo(build: str, url: str) -> None:
+    """
+    Creates all the required repo files based on the given base_url.
+
+    Args:
+        build: str          Bucket name or RH Ceph build
+        url: str       URL holding the RPMs to be uploaded
+    Returns:
+        None
+    """
+    _tmpl = Template(REPO_TEMPLATE)
+    data = dict({"nvr": build, "base_url": url})
+    repo_file = _tmpl.render(data=data)
+    LOG.debug(f"The repo file is \n {repo_file}")
+
+    with open(f"/etc/yum.repos.d/{build}.repo", "w") as fh:
+        fh.write(repo_file)
+
+    LOG.info("Successfully created repo file.")
+
+
+def compress_build(build: str) -> str:
+    """
+    Returns the compressed file name containing the development RPMs.
+
+    Args:
+        build: str  The base repo id to be used to pull the packages.
+    Returns:
+        str: The complete file path to the archive file created.
+    """
+    temp_dir = tempfile.mkdtemp()
+    _file = "".join(random.choices(string.ascii_lowercase + string.digits, k=5))
+    temp_file = join("/tmp", f"{_file}.tar.gz")
+
+    for repo in ["OSD", "MON", "Tools"]:
+        subprocess.run(
+            [
+                "reposync",
+                "-l",
+                "-d",
+                "-m",
+                "--newest-only",
+                "--download-metadata",
+                "--repoid",
+                f"{build}-{repo}",
+                "--download_path",
+                temp_dir,
+            ],
+            check=True,
+        )
+
+    subprocess.run(["tar", "-cvzf", temp_file, temp_dir], check=True)
+    LOG.info(f"Archive file successfully created - {temp_file}")
+
+    return temp_file
+
+
+def upload_object(file_name: str, bucket: str, item: str) -> None:
+    """
+    Uploads the given file to the provided bucket with the mentioned name.
+
+    Args:
+        file_name: str  Complete path to the file that needs to be uploaded
+        bucket:str      The name of container to which file has to be uploaded
+        item: str       The name to be used for the uploaded object
+    """
+    cos = CloudObjectStorage()
+
+    if not cos.bucket_exists(bucket):
```

Comment:
This type of "check before creating" is the Look Before You Leap pattern. I think it would be simpler and less racy to try create the bucket every time, and then just catch the exception if it fails in a way you don't expect.

https://stackoverflow.com/questions/26871884/how-can-i-easily-determine-if-a-boto-3-s3-bucket-resource-exists indicates that there may be a way to do this with boto that simply gives you the bucket, without raising an exception?